test/functional/forall/test-forall-rangesegment.hpp

```diff
@@ -0,0 +1,73 @@
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+// Copyright (c) 2016-20, Lawrence Livermore National Security, LLC
+// and RAJA project contributors. See the RAJA/COPYRIGHT file for details.
+//
+// SPDX-License-Identifier: (BSD-3-Clause)
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#ifndef __TEST_FORALL_RANGESEGMENT_HPP__
+#define __TEST_FORALL_RANGESEGMENT_HPP__
+
+#include "RAJA/RAJA.hpp"
+#include "camp/resource.hpp"
+
+template<typename T>
+void ForallRangeSegmentFunctionalTest_host(T first, T last)
+{
+  RAJA::TypedRangeSegment<T> r1(first, last);
+  RAJA::Index_type N = r1.end() - r1.begin();
+
+  camp::resources::Resource res{camp::resources::Host()};
+  T * ref_array = res.allocate<T>(N);
+
+  for(RAJA::Index_type i=0; i < r1.size(); i++)
+  {
+    ref_array[i]= *r1.begin() + i;
+  }
+
+  RAJA::forall<RAJA::seq_exec>(r1, [=](RAJA::Index_type idx){
+    ASSERT_EQ(ref_array[idx-*r1.begin()], idx);
+  });
+
+  res.deallocate(ref_array);
+}
+
+#if defined(RAJA_ENABLE_CUDA)
+template<typename T>
+void ForallRangeSegmentFunctionalTest_cuda(T first, T last)
```

Comment:
As an example of combining @artv3's and @davidbeckingsale's suggestions, we would want the test case call for this function to look something like:
`ForallRangeSegmentFunctionalTest<TypeParam>(first, last, camp::resources::Resource temp{camp::resources::Cuda()});`

The function signature would change to something like:
`void ForallRangeSegmentFunctionalTest(T first, T last, camp::resources::Resource working)`

And finally, `dev_res` would become `working`, and the `host_res` would become `checking`.

Reply:
I want to try building up a list of testing TypeParams which contain information on our iteration type, execution policy and resource architecture, so for now I am going to try and pass the resource type during compile time, but if that doesn't work out I'll go ahead and revert back to passing it how you've suggested like this. Thanks @rchen20 !